WalletWasabi.Fluent/ViewModels/SearchBar/Sources/TransactionsSearchSource.cs

```diff
@@ -0,0 +1,104 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Disposables;
+using System.Reactive.Linq;
+using System.Threading.Tasks;
+using DynamicData;
+using ReactiveUI;
+using WalletWasabi.Fluent.ViewModels.SearchBar.Patterns;
+using WalletWasabi.Fluent.ViewModels.SearchBar.SearchItems;
+using WalletWasabi.Fluent.ViewModels.Wallets;
+using WalletWasabi.Fluent.ViewModels.Wallets.Home.History.HistoryItems;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.Fluent.ViewModels.SearchBar.Sources;
+
+public class TransactionsSearchSource : ReactiveObject, ISearchSource, IDisposable
+{
+	private readonly CompositeDisposable _disposables = new();
+
+	public TransactionsSearchSource(IObservable<string> query)
+	{
+		var sourceCache = new SourceCache<ISearchItem, ComposedKey>(x => x.Key)
+			.DisposeWith(_disposables);
+
+		query
+			.Select(s => s.Length > 3 ? PerformSearch(s) : Enumerable.Empty<ISearchItem>())
+			.Do(results => sourceCache.Edit(e => e.Load(results)))
+			.Subscribe()
+			.DisposeWith(_disposables);
+
+		Changes = sourceCache.Connect();
+	}
+
+	public IObservable<IChangeSet<ISearchItem, ComposedKey>> Changes { get; }
+
+	private static bool Contains(string queryStr, HistoryItemViewModelBase historyItemViewModelBase)
+	{
+		return historyItemViewModelBase.Id.ToString().Contains(queryStr, StringComparison.CurrentCultureIgnoreCase);
+	}
+
+	private static Task NavigateTo(WalletViewModel wallet, HistoryItemViewModelBase item)
+	{
+		wallet.NavigateAndHighlight(item.Id);
+		return Task.CompletedTask;
+	}
+
+	private static string GetIcon(HistoryItemViewModelBase historyItemViewModelBase)
+	{
+		return historyItemViewModelBase switch
+		{
+			CoinJoinHistoryItemViewModel => "shield_regular",
+			CoinJoinsHistoryItemViewModel => "shield_regular",
+			TransactionHistoryItemViewModel => "normal_transaction",
+			_ => ""
+		};
+	}
+
+	private static IEnumerable<(WalletViewModel, HistoryItemViewModelBase)> Flatten(
+		IEnumerable<(WalletViewModel Wallet, IEnumerable<HistoryItemViewModelBase> Transactions)>
+			walletTransactions)
```

Comment:
I have a hard time reading this... maybe remove one enter? 

```
private static IEnumerable<(WalletViewModel, HistoryItemViewModelBase)> Flatten(
		IEnumerable<(WalletViewModel Wallet, IEnumerable<HistoryItemViewModelBase> Transactions)> walletTransactions)
```

Reply:
This is intrinsically hard, I'm afraid.

It transforms tuples like this:

```csharp
(Key = 1, Results = { 1, 2, 3, 4 }) 
```

into tuples like this

```csharp
(Key=1, Result = 1),
(Key=1, Result = 2),
(Key=1, Result = 3),
(Key=1, Result = 4)
```

I can't think of a better way to improve the readability :(

Reply:
I agree with @ichthus1604, the last line break is unnecessary. Please make the line breaks as @ichthus1604 suggested.

```
	private static IEnumerable<(WalletViewModel, HistoryItemViewModelBase)> Flatten(
		IEnumerable<(WalletViewModel Wallet, IEnumerable<HistoryItemViewModelBase> Transactions)> walletTransactions)
```

BTW I have generally problems with the wrapping... In the modern world where everyone has a decent resolution or even a wide screen, wrapping the lines just decreases the readability. (That is the first thing that I usually turn off in an IDE)